lib/secretsManager.js

```diff
@@ -1,77 +1,169 @@
- * When provided, all where clause must be fulfilled (logical AND).
- * By default, returns all secrets
- * @async
- * @param {Object} where - hash of where clauses returned secrets must fulfill with
- * @returns {Array<Object>} list (may be empty) of matching secrets
- */
-export async function find(where = {}) {
-  let secrets = []
-
-  try {
-    const props = Object.keys(where)
-    secrets = DUMMY_STORE.filter(secret =>
-      props.every(prop => secret[prop] === where[prop])
+import { v4 as uuid } from 'uuid'
+import aesjs from 'aes-js'
+import * as SecureStore from 'expo-secure-store'
+import AsyncStorage from '@react-native-async-storage/async-storage'
+
+const encryptionPassphraseStorageKey = 'encryptionPassphrase'
+const secretStorageKey = 'secrets'
+
+export default class SecretsManager {
+  constructor() {
+    this.encryptionPassphrase = null
+  }
+
+  async initialize() {
+    let getEncryptionKey = AsyncStorage.getItem
+    let setEncryptionKey = AsyncStorage.setItem
+
+    if (SecureStore.isAvailableAsync()) {
+      getEncryptionKey = SecureStore.getItemAsync
+      setEncryptionKey = SecureStore.setItemAsync
+    }
+
+    const encryptionPassphrase = await getEncryptionKey(
+      encryptionPassphraseStorageKey
+    )
+
+    if (encryptionPassphrase) {
+      this.encryptionPassphrase = encryptionPassphrase
+      return
+    }
+
+    this.encryptionPassphrase = this.generateEncryptionPassphrase()
+
+    await this.flush()
+
+    await setEncryptionKey(
+      encryptionPassphraseStorageKey,
+      this.encryptionPassphrase
     )
-  } catch (err) {
-    console.error('Failed to read secrets from data store', err)
   }
 
-  return secrets
-}
+  generateEncryptionPassphrase() {
+    return JSON.stringify(
+      Array.from(
+        crypto.getRandomValues(
+          Uint8Array.from(Array.from({ length: 32, value: 0 }))
+        )
+      )
+    )
+  }
+
+  async flush() {
+    if (SecureStore.isAvailableAsync()) {
+      for (const key of [encryptionPassphraseStorageKey, secretStorageKey]) {
+        await SecureStore.deleteItemAsync(key)
+      }
+      return
+    }
+
+    await AsyncStorage.multiRemove([
+      encryptionPassphraseStorageKey,
+      secretStorageKey,
+    ])
+  }
 
-/**
- * Add or update a secret to the list.
- * If not provided, an _id key will be generated
- * @async
- * @param {Object} secret - secret to upsert
- * @returns {Object} the added/updated secret
- */
-export async function upsert(secret) {
-  const secretId = secret._id || DUMMY_UUID_GEN++
-  const secrets = await find()
-  let upserted = secrets.find(({ _id }) => _id === secretId)
-
-  if (upserted) {
-    Object.assign(upserted, secret)
-  } else {
-    upserted = { _id: secretId, ...secret }
-    secrets.push(upserted)
+  async encrypt(secrets) {
+    const secretsBytes = aesjs.utils.utf8.toBytes(JSON.stringify(secrets))
+    const aesCtr = new aesjs.ModeOfOperation.ctr(
+      Uint8Array.from(JSON.parse(this.encryptionPassphrase)),
```

Comment:
why is encryptionPassphrase a json string? I think we should parse it to a JavaScript value when we initialize/read it, so that the rest of the code in this class doesn't have to deal with JSON

Reply:
I'll see what I can do about it. I'm not sure changing it will give us any more readable or maintainable alternative. Thanks for bringing it up anyway.

Reply:
I've changed it per your request after all. 👍